src/security/Certificate.cc

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "debug/Stream.h"
+#include "sbuf/SBuf.h"
+#include "security/Certificate.h"
+
+#if USE_OPENSSL
+#include "ssl/gadgets.h"
+#endif
+
+#include <iostream>
+
+inline
+const char *
+MissingLibraryError()
+{
+    return "[need OpenSSL or GnuTLS]";
+}
+
+SBuf
+Security::IssuerName(Certificate &cert)
+{
+    SBuf out;
+
+#if USE_OPENSSL
+    Ssl::ForgetErrors();
+    const auto name = X509_NAME_oneline(X509_get_issuer_name(&cert), nullptr, 0);
+    if (!name) {
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate Issuer:" <<
+               Ssl::ReportAndForgetErrors);
+        return out;
+    }
+    out.append(name);
+    OPENSSL_free(name);
+
+#elif USE_GNUTLS
+    gnutls_x509_dn_t issuer;
+    auto x = gnutls_x509_crt_get_issuer(&cert, &issuer);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate Issuer: " << ErrorString(x));
+        return out;
+    }
+
+    gnutls_datum_t name;
+    x = gnutls_x509_dn_get_str(issuer, &name);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot describe certificate Issuer: " << ErrorString(x));
+        return out;
+    }
+    out.append(reinterpret_cast<const char *>(name.data), name.size);
+    gnutls_free(name.data);
+
+#else
+    debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate Issuer: " << MissingLibraryError());
+#endif
+
+    return out;
+}
+
+SBuf
+Security::SubjectName(Certificate &cert)
+{
+    SBuf out;
+
+#if USE_OPENSSL
+    Ssl::ForgetErrors();
+    const auto name = X509_NAME_oneline(X509_get_subject_name(&cert), nullptr, 0);
+    if (!name) {
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate SubjectName:" <<
+               Ssl::ReportAndForgetErrors);
+        return out;
+    }
+    out.append(name);
+    OPENSSL_free(name);
+
+#elif USE_GNUTLS
+    gnutls_x509_dn_t subject;
+    auto x = gnutls_x509_crt_get_subject(&cert, &subject);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate SubjectName: " << ErrorString(x));
+        return out;
+    }
+
+    gnutls_datum_t name;
+    x = gnutls_x509_dn_get_str(subject, &name);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot describe certificate SubjectName: " << ErrorString(x));
+        return out;
+    }
+    out.append(reinterpret_cast<const char *>(name.data), name.size);
+    gnutls_free(name.data);
```

Comment:
The above and similar PR code is not exception-safe -- if out.append() throws, we are going to leak `name` memory. I think this should be fixed, but I am not sure whether this should be done here or in a dedicated PR: There are similar leaks in the official code, in the PR from where this code was copied, and properly addressing these leaks will require non-trivial out-of-scope changes (to make the solutions easier to reuse). I lean towards ignoring this (very unlikely!) leak in this PR and waiting for a proper reusable fix.

Reply:
I still think this is far easier than you make out. I do not insist on this.